nlmod/util.py

```diff
@@ -89,29 +94,276 @@ def get_model_dirs(model_ws):
+            download_if_not_found=download_if_not_found,
+            version_tag=version_tag,
+            repo=repo,
+            disable_version_check=disable_version_check,
+        )
+        / exe_name
+    )
+
+    msg = f"Executable path: {exe_full_path}"
+    logger.debug(msg)
+
+    return exe_full_path
+
+
+def get_bin_directory(
+    bindir=None,
+    exe_name="mf6",
+    download_if_not_found=True,
+    version_tag="latest",
+    repo="executables",
+    disable_version_check=False,
+) -> Path:
+    """
+    Get the directory where the executables are stored.
 
-    if not os.path.exists(exe_path):
-        logger.warning(
-            f"executable {exe_path} not found, download the binaries using nlmod.util.download_mfbinaries"
+    Searching for the executables is done in the following order:
+    1. The directory specified with `bindir`. Raises error if exe_name is provided
+        and not found. Requires disable_version_check to be False.
+    2. The directory used by nlmod installed in this environment.
+    3. If the executables were downloaded with flopy/nlmod from an other env,
+        most recent installation location of MODFLOW is found in flopy metadata
+
+    Else:
+    4. Download the executables using `version_tag` and `repo`.
+
+    The returned directory is checked to contain exe_name if exe_name is provided. If exe_name
+    is set to None only the existence of the directory is checked.
+
+    Parameters
+    ----------
+    bindir : Path, optional
+        The directory where the executables are stored, by default "mf6".
+    exe_name : str, optional
+        The name of the executable, by default None.
+    download_if_not_found : bool, optional
+        Download the executables if they are not found, by default True.
+    repo : str, default "executables"
+        Name of GitHub repository. Choose one of "executables" (default),
+        "modflow6", or "modflow6-nightly-build". Used only if download is needed.
+    version_tag : str, default "latest"
+        GitHub release ID. Used only if download is needed.
+    disable_version_check : bool, default False
+        If False, the most recent installation location of MODFLOW is found in flopy metadata
+        that respects `version_tag` and `repo`.
+
+    Returns
+    -------
+    Path
+        The directory where the executables are stored.
+
+    Raises
+    ------
+    FileNotFoundError
+        If the executables are not found in the specified directories.
+    """
+    bindir = Path(bindir) if bindir is not None else None
+
+    if sys.platform.startswith("win") and not exe_name.endswith(".exe"):
+        exe_name += ".exe"
+
+    # If bindir is provided
+    if bindir is not None and not disable_version_check:
+        msg = "Incompatible arguments. If bindir is provided, disable_version_check should be True."
+        raise ValueError(msg)
+
+    use_bindir = (
+        bindir is not None and exe_name is not None and Path(bindir / exe_name).exists()
```

Comment:
Don't think you need to wrap `bindir/ exe_name` in another Path call, it is already a pathlib.Path I think?